<a href="https://colab.research.google.com/github/astrovishalthakur/MachineLearning/blob/main/NaturalLanguageProcessing/Basics/TextPreprocessing.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

### NLP PIPELINE

We will se how to do basic text processing as a nlp pipeline

In [1]:
import requests

In [2]:
url1 = "https://api.themoviedb.org/3/movie/top_rated?api_key=73326928ce817d080bf46bdc0bc675b2&language=en-US&page="

In [3]:
url2 = "https://api.themoviedb.org/3/genre/movie/list?api_key=73326928ce817d080bf46bdc0bc675b2&language=en-US"

In [4]:
details = []

In [5]:
data = requests.get(f'{url1}1')

In [6]:
pages = data.json()["total_pages"]

In [7]:
type(data.json()["total_pages"])

int

In [8]:
type(data.json()["results"])

list

In [9]:
for i in range(1, pages+1):
  data = requests.get(f"{url1}{i}")
  details.extend(data.json()["results"])

In [10]:
len(details)

9736

In [11]:
details[1]

{'adult': False,
 'backdrop_path': '/5hNcsnMkwU2LknLoru73c76el3z.jpg',
 'genre_ids': [35, 18, 10749],
 'id': 19404,
 'original_language': 'hi',
 'original_title': 'दिलवाले दुल्हनिया ले जायेंगे',
 'overview': 'Raj is a rich, carefree, happy-go-lucky second generation NRI. Simran is the daughter of Chaudhary Baldev Singh, who in spite of being an NRI is very strict about adherence to Indian values. Simran has left for India to be married to her childhood fiancé. Raj leaves for India with a mission at his hands, to claim his lady love under the noses of her whole family. Thus begins a saga.',
 'popularity': 30.063,
 'poster_path': '/2CAL2433ZeIihfX1Hb2139CX0pW.jpg',
 'release_date': '1995-10-20',
 'title': 'Dilwale Dulhania Le Jayenge',
 'video': False,
 'vote_average': 8.7,
 'vote_count': 3475}

In [12]:
genres = requests.get(url2)
ids = genres.json()

In [13]:
ids

{'genres': [{'id': 28, 'name': 'Action'},
  {'id': 12, 'name': 'Adventure'},
  {'id': 16, 'name': 'Animation'},
  {'id': 35, 'name': 'Comedy'},
  {'id': 80, 'name': 'Crime'},
  {'id': 99, 'name': 'Documentary'},
  {'id': 18, 'name': 'Drama'},
  {'id': 10751, 'name': 'Family'},
  {'id': 14, 'name': 'Fantasy'},
  {'id': 36, 'name': 'History'},
  {'id': 27, 'name': 'Horror'},
  {'id': 10402, 'name': 'Music'},
  {'id': 9648, 'name': 'Mystery'},
  {'id': 10749, 'name': 'Romance'},
  {'id': 878, 'name': 'Science Fiction'},
  {'id': 10770, 'name': 'TV Movie'},
  {'id': 53, 'name': 'Thriller'},
  {'id': 10752, 'name': 'War'},
  {'id': 37, 'name': 'Western'}]}

In [14]:
new_ids = {}
for items in ids["genres"]:
  new_ids[items["id"]]= items["name"]

In [15]:
new_ids

{12: 'Adventure',
 14: 'Fantasy',
 16: 'Animation',
 18: 'Drama',
 27: 'Horror',
 28: 'Action',
 35: 'Comedy',
 36: 'History',
 37: 'Western',
 53: 'Thriller',
 80: 'Crime',
 99: 'Documentary',
 878: 'Science Fiction',
 9648: 'Mystery',
 10402: 'Music',
 10749: 'Romance',
 10751: 'Family',
 10752: 'War',
 10770: 'TV Movie'}

In [16]:
finalLis = []
dic = {}
for item in details:
  dic = {"title":item["title"],
         "genres": [new_ids[i] for i in item["genre_ids"]],
         "description":item["overview"]}
  finalLis.append(dic)
  dic = {}

In [17]:
finalLis[:5]

[{'description': 'Framed in the 1940s for the double murder of his wife and her lover, upstanding banker Andy Dufresne begins a new life at the Shawshank prison, where he puts his accounting skills to work for an amoral warden. During his long stretch in prison, Dufresne comes to be admired by the other inmates -- including an older prisoner named Red -- for his integrity and unquenchable sense of hope.',
  'genres': ['Drama', 'Crime'],
  'title': 'The Shawshank Redemption'},
 {'description': 'Raj is a rich, carefree, happy-go-lucky second generation NRI. Simran is the daughter of Chaudhary Baldev Singh, who in spite of being an NRI is very strict about adherence to Indian values. Simran has left for India to be married to her childhood fiancé. Raj leaves for India with a mission at his hands, to claim his lady love under the noses of her whole family. Thus begins a saga.',
  'genres': ['Comedy', 'Drama', 'Romance'],
  'title': 'Dilwale Dulhania Le Jayenge'},
 {'description': 'Spanning

In [18]:
import pandas as pd

In [19]:
df = pd.DataFrame(finalLis, columns=["description", "genres", "title"])

In [20]:
df.head()

description  \
0  Framed in the 1940s for the double murder of h...   
1  Raj is a rich, carefree, happy-go-lucky second...   
2  Spanning the years 1945 to 1955, a chronicle o...   
3  A tragic accident lead to Kaori's blindness, b...   
4  The true story of how businessman Oskar Schind...   

                     genres                        title  
0            [Drama, Crime]     The Shawshank Redemption  
1  [Comedy, Drama, Romance]  Dilwale Dulhania Le Jayenge  
2            [Drama, Crime]                The Godfather  
3          [Romance, Drama]               Your Eyes Tell  
4     [Drama, History, War]             Schindler's List

In [21]:
df = df[["title", "genres", "description"]]

In [22]:
df

title  \
0             The Shawshank Redemption   
1          Dilwale Dulhania Le Jayenge   
2                        The Godfather   
3                       Your Eyes Tell   
4                     Schindler's List   
...                                ...   
9731                 Alone in the Dark   
9732                 House of the Dead   
9733              Dragonball Evolution   
9734  Box Office 3D - Il film dei film   
9735        Birdemic: Shock and Terror   

                                                 genres  \
0                                        [Drama, Crime]   
1                              [Comedy, Drama, Romance]   
2                                        [Drama, Crime]   
3                                      [Romance, Drama]   
4                                 [Drama, History, War]   
...                                                 ...   
9731                          [Action, Fantasy, Horror]   
9732                         [Horror, Action, Thriller]   
9733  [Action, Adventure, Fantasy, Science Fiction, ...   
9734                                           [Comedy]   
9735       [Science Fiction, Horror, Romance, Thriller]   

                                            description  
0     Framed in the 1940s for the double murder of h...  
1     Raj is a rich, carefree, happy-go-lucky second...  
2     Spanning the years 1945 to 1955, a chronicle o...  
3     A tragic accident lead to Kaori's blindness, b...  
4     The true story of how businessman Oskar Schind...  
...                                                 ...  
9731  Edward Carnby is a private investigator specia...  
9732  Set on an island off the coast, a techno rave ...  
9733  The young warrior Son Goku sets out on a quest...  
9734  An Italian parody of the biggest U.S. blockbus...  
9735  A platoon of eagles and vultures attacks the r...  

[9736 rows x 3 columns]

# Making everything Lower Case

In [23]:
df.description = df.description.str.lower()

In [24]:
df.description

0       framed in the 1940s for the double murder of h...
1       raj is a rich, carefree, happy-go-lucky second...
2       spanning the years 1945 to 1955, a chronicle o...
3       a tragic accident lead to kaori's blindness, b...
4       the true story of how businessman oskar schind...
                              ...                        
9731    edward carnby is a private investigator specia...
9732    set on an island off the coast, a techno rave ...
9733    the young warrior son goku sets out on a quest...
9734    an italian parody of the biggest u.s. blockbus...
9735    a platoon of eagles and vultures attacks the r...
Name: description, Length: 9736, dtype: object

# Removing HTML Tags

example

In [25]:
import re
def striphtml(data):
    p = re.compile(r'<.*?>')
    return p.sub('', data)

striphtml('<a href="foo.com" class="bar">I Want This <b>text!</b></a>')


'I Want This text!'

In [26]:
k = df.description.apply(lambda x:striphtml(x))

In [27]:
for item1, item2 in zip(k, df.description):
  if item1 != item2:
    print(item2)

In [28]:
dic = dict(k == df.description)

In [29]:
type(dic)

dict

In [30]:
for i in dic:
  if dic[i] == False:
    print(dic[i])

since above code gave no output, there were no html tags in our description

Looks like we don't need to remove html tags

let's make sure with this example that above approach is right to determine whether we want to remove html tags.

In [31]:
sr = pd.Series(["hello", "my name is <p>vishal</p>thakur", "Who <html>are</html>you"])

In [32]:
sr

0                             hello
1    my name is <p>vishal</p>thakur
2           Who <html>are</html>you
dtype: object

In [33]:
kk = sr.apply(lambda x: striphtml(x))

In [34]:
kk

0                      hello
1    my name is vishalthakur
2                 Who areyou
dtype: object

In [35]:
dic2 = dict(kk == sr)

In [36]:
dic2

{0: True, 1: False, 2: False}

In [37]:
for i in dic2:
  if dic2[i] == False:
    print(i)

1
2


as you can see, above code gave output of 1 and 2, but same code didn't gave any output on our description.

# Removing Puntuations

In [38]:
pattern = re.compile(r'[^\w\s]')

In [39]:
re.sub(pattern,"", "hello! my name is ? vishal")

'hello my name is  vishal'

In [40]:
df.description = df.description.apply(lambda x: re.sub(pattern, "", x))

In [41]:
df.description

0       framed in the 1940s for the double murder of h...
1       raj is a rich carefree happygolucky second gen...
2       spanning the years 1945 to 1955 a chronicle of...
3       a tragic accident lead to kaoris blindness but...
4       the true story of how businessman oskar schind...
                              ...                        
9731    edward carnby is a private investigator specia...
9732    set on an island off the coast a techno rave p...
9733    the young warrior son goku sets out on a quest...
9734    an italian parody of the biggest us blockbuste...
9735    a platoon of eagles and vultures attacks the r...
Name: description, Length: 9736, dtype: object

In [42]:
df

title  \
0             The Shawshank Redemption   
1          Dilwale Dulhania Le Jayenge   
2                        The Godfather   
3                       Your Eyes Tell   
4                     Schindler's List   
...                                ...   
9731                 Alone in the Dark   
9732                 House of the Dead   
9733              Dragonball Evolution   
9734  Box Office 3D - Il film dei film   
9735        Birdemic: Shock and Terror   

                                                 genres  \
0                                        [Drama, Crime]   
1                              [Comedy, Drama, Romance]   
2                                        [Drama, Crime]   
3                                      [Romance, Drama]   
4                                 [Drama, History, War]   
...                                                 ...   
9731                          [Action, Fantasy, Horror]   
9732                         [Horror, Action, Thriller]   
9733  [Action, Adventure, Fantasy, Science Fiction, ...   
9734                                           [Comedy]   
9735       [Science Fiction, Horror, Romance, Thriller]   

                                            description  
0     framed in the 1940s for the double murder of h...  
1     raj is a rich carefree happygolucky second gen...  
2     spanning the years 1945 to 1955 a chronicle of...  
3     a tragic accident lead to kaoris blindness but...  
4     the true story of how businessman oskar schind...  
...                                                 ...  
9731  edward carnby is a private investigator specia...  
9732  set on an island off the coast a techno rave p...  
9733  the young warrior son goku sets out on a quest...  
9734  an italian parody of the biggest us blockbuste...  
9735  a platoon of eagles and vultures attacks the r...  

[9736 rows x 3 columns]

# Spelling correction

In [43]:
! pip install textblob

In [44]:
from textblob import TextBlob

In [45]:
str1 = "helo, my nami is vishal Thakur. Who are you"

In [46]:
blb = TextBlob(str1)

In [47]:
blb.correct().string

'held, my name is visual Thakur. Who are you'

looks like it is making some mistakes on non english words. Also, since this is movied descriptions, it will not have many spelling mistakes. So let's not do any spell corrections

# Stop words treatment

In [48]:
! pip install nltk

In [49]:
import nltk

In [50]:
nltk.download("stopwords")
from nltk.corpus import stopwords

[nltk_data] Downloading package stopwords to /root/nltk_data...
[nltk_data]   Unzipping corpora/stopwords.zip.


In [51]:
stopwords.words("english")[:10]

['i', 'me', 'my', 'myself', 'we', 'our', 'ours', 'ourselves', 'you', "you're"]

In [52]:
len(stopwords.words("english"))

179

looks like we have 179 stop words

We will split the text, then remove stop words.

In [53]:
# let's see how join works
a = ["a", "b", "c", "d"]
" ".join(a)

'a b c d'

In [54]:
def remove_stop(text):
  lis = []
  for word in text.split():
    if word not in stopwords.words("english"):
      lis.append(word)
  return " ".join(lis)


as you can see, we are joining every thing back. We will be tokenizing everything again, so we don't need to join it back. We can just return the list.

`return lis`

In [55]:
series = df.description.apply(lambda x: remove_stop(x))

In [56]:
df.description = series

In [57]:
df

title  \
0             The Shawshank Redemption   
1          Dilwale Dulhania Le Jayenge   
2                        The Godfather   
3                       Your Eyes Tell   
4                     Schindler's List   
...                                ...   
9731                 Alone in the Dark   
9732                 House of the Dead   
9733              Dragonball Evolution   
9734  Box Office 3D - Il film dei film   
9735        Birdemic: Shock and Terror   

                                                 genres  \
0                                        [Drama, Crime]   
1                              [Comedy, Drama, Romance]   
2                                        [Drama, Crime]   
3                                      [Romance, Drama]   
4                                 [Drama, History, War]   
...                                                 ...   
9731                          [Action, Fantasy, Horror]   
9732                         [Horror, Action, Thriller]   
9733  [Action, Adventure, Fantasy, Science Fiction, ...   
9734                                           [Comedy]   
9735       [Science Fiction, Horror, Romance, Thriller]   

                                            description  
0     framed 1940s double murder wife lover upstandi...  
1     raj rich carefree happygolucky second generati...  
2     spanning years 1945 1955 chronicle fictional i...  
3     tragic accident lead kaoris blindness clings l...  
4     true story businessman oskar schindler saved t...  
...                                                 ...  
9731  edward carnby private investigator specializin...  
9732  set island coast techno rave party attracts di...  
9733  young warrior son goku sets quest racing time ...  
9734  italian parody biggest us blockbusters gladiat...  
9735  platoon eagles vultures attacks residents smal...  

[9736 rows x 3 columns]

# Tokenization
It's time to tokenize

In [58]:
from nltk.tokenize import word_tokenize
nltk.download("punkt")

[nltk_data] Downloading package punkt to /root/nltk_data...
[nltk_data]   Unzipping tokenizers/punkt.zip.


True

In [59]:
word_tokenize("Hello my name is vishal thakur.")

['Hello', 'my', 'name', 'is', 'vishal', 'thakur', '.']

In [60]:
df.description.iloc[0]

'framed 1940s double murder wife lover upstanding banker andy dufresne begins new life shawshank prison puts accounting skills work amoral warden long stretch prison dufresne comes admired inmates including older prisoner named red integrity unquenchable sense hope'

Checking on one example

In [61]:
word_tokenize(df.description.iloc[0])

['framed',
 '1940s',
 'double',
 'murder',
 'wife',
 'lover',
 'upstanding',
 'banker',
 'andy',
 'dufresne',
 'begins',
 'new',
 'life',
 'shawshank',
 'prison',
 'puts',
 'accounting',
 'skills',
 'work',
 'amoral',
 'warden',
 'long',
 'stretch',
 'prison',
 'dufresne',
 'comes',
 'admired',
 'inmates',
 'including',
 'older',
 'prisoner',
 'named',
 'red',
 'integrity',
 'unquenchable',
 'sense',
 'hope']

Applying to whole description column of dataset

In [62]:
df.description = df.description.apply(lambda x: word_tokenize(x))
df.description

0       [framed, 1940s, double, murder, wife, lover, u...
1       [raj, rich, carefree, happygolucky, second, ge...
2       [spanning, years, 1945, 1955, chronicle, ficti...
3       [tragic, accident, lead, kaoris, blindness, cl...
4       [true, story, businessman, oskar, schindler, s...
                              ...                        
9731    [edward, carnby, private, investigator, specia...
9732    [set, island, coast, techno, rave, party, attr...
9733    [young, warrior, son, goku, sets, quest, racin...
9734    [italian, parody, biggest, us, blockbusters, g...
9735    [platoon, eagles, vultures, attacks, residents...
Name: description, Length: 9736, dtype: object

# Stemming

In [63]:
# import these modules
from nltk.stem import PorterStemmer
from nltk.tokenize import word_tokenize

ps = PorterStemmer()

# choose some words to be stemmed
words = ["program", "programs", "programmer", "programming", "programmers"]

for w in words:
	print(w, " : ", ps.stem(w))


program  :  program
programs  :  program
programmer  :  programm
programming  :  program
programmers  :  programm


In [64]:
from nltk.stem import WordNetLemmatizer
wordnet = WordNetLemmatizer()

In [65]:
nltk.download("wordnet")

[nltk_data] Downloading package wordnet to /root/nltk_data...
[nltk_data]   Unzipping corpora/wordnet.zip.


True

# Learning Formatted Output

That's how we align things.

In [66]:

# here 20 spaces are reserved for the 
# particular output string. And the string
# is printed on the left side
print(f"{'Left Aligned Text' : <20}")

Left Aligned Text   


In [67]:

# here 20 spaces are reserved for the 
# particular output string. And the string
# is printed on the right side
print(f"{'Right Aligned Text' : >20}")

  Right Aligned Text


In [68]:
# here 20 spaces are reserved for the 
# particular output string. And the string
# is printed in the middle
print(f"{'Centered' : ^10}")

 Centered 


In [69]:
# assigning strings to the variables
left_alignment = "Left Text"
center_alignment = "Centered Text"
right_alignment = "Right Text"

# printing out aligned text
print(f"{left_alignment : <20}{center_alignment : ^15}{right_alignment : >20}")


Left Text            Centered Text           Right Text


In [70]:

# assigning list values to the variables
names = ['Raj', 'Shivam', 'Shreeya', 'Kartik']
marks = [7, 9, 8, 5]
div = ['A', 'A', 'C', 'B']
id = [21, 52, 27, 38]


print(f"{'Name' : <10}{'Marks' : ^10}{'Division' : ^10}{'ID' : >5}")
  
# printing values of variables in Aligned manner
for i in range(0, 4):
    print(f"{names[i] : <10}{marks[i] : ^10}{div[i] : ^10}{id[i] : >5}")

Name        Marks    Division    ID
Raj           7         A        21
Shivam        9         A        52
Shreeya       8         C        27
Kartik        5         B        38


In [115]:
print(f"{'Original':<20}{'Stemmed':^20}{'Lemmatized':^20}")
for w in df.description[6]:
  print(f"{w:<20}{ps.stem(w):^20}{wordnet.lemmatize(w):^20}")

Original                  Stemmed            Lemmatized     
continu                   continu             continu       
saga                        saga                saga        
corleon                   corleon             corleon       
crime                      crime               crime        
famili                     famili              famili       
young                      young               young        
vito                        vito                vito        
corleon                   corleon             corleon       
grow                        grow                grow        
sicili                     sicili              sicili       
1910                        1910                1910        
new                         new                 new         
york                        york                york        
1950                        1950                1950        
michael                   michael             michael       
corleon                 

looks like both are giving good results, but we will manually check for some words

In [72]:
a = ["continue", "continued", "countinuing", "countinous", "continuing"]

for w in a:
  print(w, "  :  ", ps.stem(w), "  :  ", wordnet.lemmatize(w))

continue   :   continu   :   continue
continued   :   continu   :   continued
countinuing   :   countinu   :   countinuing
countinous   :   countin   :   countinous
continuing   :   continu   :   continuing


since lemmatization gives so many version above words, we will go with stemming

In [73]:
des_stemmed = df.description.apply(lambda x: [ps.stem(i) for i in x])

In [74]:
des_stemmed

0       [frame, 1940, doubl, murder, wife, lover, upst...
1       [raj, rich, carefre, happygolucki, second, gen...
2       [span, year, 1945, 1955, chronicl, fiction, it...
3       [tragic, accid, lead, kaori, blind, cling, lif...
4       [true, stori, businessman, oskar, schindler, s...
                              ...                        
9731    [edward, carnbi, privat, investig, special, un...
9732    [set, island, coast, techno, rave, parti, attr...
9733    [young, warrior, son, goku, set, quest, race, ...
9734    [italian, parodi, biggest, us, blockbust, glad...
9735    [platoon, eagl, vultur, attack, resid, small, ...
Name: description, Length: 9736, dtype: object

In [75]:
df.description = des_stemmed

# And here we have our dataset, ready to be modelled upon.

In [76]:
df

title  \
0             The Shawshank Redemption   
1          Dilwale Dulhania Le Jayenge   
2                        The Godfather   
3                       Your Eyes Tell   
4                     Schindler's List   
...                                ...   
9731                 Alone in the Dark   
9732                 House of the Dead   
9733              Dragonball Evolution   
9734  Box Office 3D - Il film dei film   
9735        Birdemic: Shock and Terror   

                                                 genres  \
0                                        [Drama, Crime]   
1                              [Comedy, Drama, Romance]   
2                                        [Drama, Crime]   
3                                      [Romance, Drama]   
4                                 [Drama, History, War]   
...                                                 ...   
9731                          [Action, Fantasy, Horror]   
9732                         [Horror, Action, Thriller]   
9733  [Action, Adventure, Fantasy, Science Fiction, ...   
9734                                           [Comedy]   
9735       [Science Fiction, Horror, Romance, Thriller]   

                                            description  
0     [frame, 1940, doubl, murder, wife, lover, upst...  
1     [raj, rich, carefre, happygolucki, second, gen...  
2     [span, year, 1945, 1955, chronicl, fiction, it...  
3     [tragic, accid, lead, kaori, blind, cling, lif...  
4     [true, stori, businessman, oskar, schindler, s...  
...                                                 ...  
9731  [edward, carnbi, privat, investig, special, un...  
9732  [set, island, coast, techno, rave, parti, attr...  
9733  [young, warrior, son, goku, set, quest, race, ...  
9734  [italian, parodi, biggest, us, blockbust, glad...  
9735  [platoon, eagl, vultur, attack, resid, small, ...  

[9736 rows x 3 columns]

# Featrue Extraction

## TF-IDF

In [77]:
from sklearn.feature_extraction.text import TfidfVectorizer
tfidf = TfidfVectorizer()

In [79]:
TFDF = tfidf.fit_transform(df.description.apply(lambda x: " ".join(x))).toarray()

In [80]:
TFDF

array([[0., 0., 0., ..., 0., 0., 0.],
       [0., 0., 0., ..., 0., 0., 0.],
       [0., 0., 0., ..., 0., 0., 0.],
       ...,
       [0., 0., 0., ..., 0., 0., 0.],
       [0., 0., 0., ..., 0., 0., 0.],
       [0., 0., 0., ..., 0., 0., 0.]])

In [81]:
print(tfidf.idf_)
print(tfidf.get_feature_names_out())

[9.08507605 9.08507605 8.23777819 ... 9.49054116 9.49054116 9.49054116]
['00' '000' '007' ... 'öztürk' 'fight' 'first']


In [82]:
from sklearn.feature_extraction.text import CountVectorizer
cv = CountVectorizer()

In [84]:
bow = cv.fit_transform(df.description.apply(lambda x: " ".join(x)))

In [106]:
type(cv.vocabulary_)

dict

In [111]:
n = 1
for i, j in cv.vocabulary_.items():
  print(f"{i} : {j}")
  n += 1
  if n > 20:
    break

frame : 7817
1940 : 193
doubl : 6016
murder : 13615
wife : 22102
lover : 12079
upstand : 21296
banker : 1992
andi : 1209
dufresn : 6209
begin : 2236
new : 13933
life : 11752
shawshank : 18076
prison : 15835
put : 16070
account : 626
skill : 18426
work : 22280
amor : 1156


In [87]:
bow.shape

(9736, 22649)

## CountVectorizer----BAG OF WORDS

In [88]:
cv2 = CountVectorizer(ngram_range=(2, 3))

In [90]:
bow2 = cv2.fit_transform(df.description.apply(lambda x:" ".join(x)))

In [99]:
print(len(cv2.vocabulary_))

439270


In [101]:
import time

In [102]:
for i in cv2.vocabulary_:
  time.sleep(1)
  print(i)

frame 1940
1940 doubl
doubl murder
murder wife
wife lover
lover upstand
upstand banker
banker andi
andi dufresn
dufresn begin
begin new
new life
life shawshank
shawshank prison
prison put
put account
account skill
skill work
work amor
amor warden
warden long
long stretch
stretch prison
prison dufresn
dufresn come
come admir
admir inmat
inmat includ
includ older
older prison
prison name
name red
red integr
integr unquench
unquench sens
sens hope
frame 1940 doubl
1940 doubl murder
doubl murder wife
murder wife lover
wife lover upstand
lover upstand banker
upstand banker andi
banker andi dufresn
andi dufresn begin
dufresn begin new
begin new life
new life shawshank
life shawshank prison
shawshank prison put
prison put account
put account skill
account skill work
skill work amor
work amor warden
amor warden long
warden long stretch
long stretch prison
stretch prison dufresn
prison dufresn come
dufresn come admir
come admir inmat
admir inmat includ
inmat includ older
includ older prison
old

KeyboardInterrupt: ignored